### ISSUE TITLE
TFX example code, Exception: It appears that you are attempting to reference SparkContext from a broadcast variable, action, or transformation. SparkContext can only be used on the driver, not in code that it run on workers. For more information, see SPARK-5063.

### ISSUE THREAD

**Issue by umusa** [tensorflow/tfx#2220]
I trying to run the example code at https://www.tensorflow.org/tfx/tutorials/transform/census
on databricks GPU cluster.

My env:

     7.1 ML Spark 3.0.0 Scala 2.12 GPU
     python 3.7
     tensorflow: Version: 2.1.1
     tensorflow-transform==0.22.0
     apache_beam==2.21.0

When I run
 
      transform_data(train, test, temp)

I got error:

       Exception: It appears that you are attempting to reference SparkContext from a broadcast variable, action, or transformation. SparkContext can only be used on the driver, not in code that it run on workers. For more information, see SPARK-5063.

It seems that the worker tries to call SparkContext pickled by pyspark ? 
I am not sure how to avoid this by changing the example code ?

thanks

 the callstack 

---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
/databricks/python/lib/python3.7/site-packages/apache_beam/internal/pickler.py in dumps(o, enable_trace)
    253     try:
--> 254       s = dill.dumps(o)
    255     except Exception:  # pylint: disable=broad-except

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in dumps(obj, protocol, byref, fmode, recurse, **kwds)
    264     file = StringIO()
--> 265     dump(obj, file, protocol, byref, fmode, recurse, **kwds)#, strictio)
    266     return file.getvalue()

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in dump(obj, file, protocol, byref, fmode, recurse, **kwds)
    258     _kwds.update(dict(byref=byref, fmode=fmode, recurse=recurse))
--> 259     Pickler(file, protocol, **_kwds).dump(obj)
    260     return

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in dump(self, obj)
    444         else:
--> 445             StockPickler.dump(self, obj)
    446         stack.clear()  # clear record of 'recursion-sensitive' pickled objects

/databricks/python/lib/python3.7/pickle.py in dump(self, obj)
    436             self.framer.start_framing()
--> 437         self.save(obj)
    438         self.write(STOP)

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    548         # Save the reduce() output and finally memoize the object
--> 549         self.save_reduce(obj=obj, *rv)
    550 

/databricks/python/lib/python3.7/pickle.py in save_reduce(self, func, args, state, listitems, dictitems, obj)
    661         if state is not None:
--> 662             save(state)
    663             write(BUILD)

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return

/databricks/python/lib/python3.7/site-packages/apache_beam/internal/pickler.py in new_save_module_dict(pickler, obj)
    220     else:
--> 221       return old_save_module_dict(pickler, obj)
    222 

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in save_module_dict(pickler, obj)
    911             pickler._session = False
--> 912         StockPickler.save_dict(pickler, obj)
    913         log.info("# D2")

/databricks/python/lib/python3.7/pickle.py in save_dict(self, obj)
    858         self.memoize(obj)
--> 859         self._batch_setitems(obj.items())
    860 

/databricks/python/lib/python3.7/pickle.py in _batch_setitems(self, items)
    884                     save(k)
--> 885                     save(v)
    886                 write(SETITEMS)

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in save_function(pickler, obj)
   1412                                 obj.__defaults__, obj.__closure__,
-> 1413                                 obj.__dict__, fkwdefaults), obj=obj)
   1414         else:

/databricks/python/lib/python3.7/pickle.py in save_reduce(self, func, args, state, listitems, dictitems, obj)
    637             save(func)
--> 638             save(args)
    639             write(REDUCE)

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return

/databricks/python/lib/python3.7/pickle.py in save_tuple(self, obj)
    788         for element in obj:
--> 789             save(element)
    790 

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return

/databricks/python/lib/python3.7/pickle.py in save_tuple(self, obj)
    773             for element in obj:
--> 774                 save(element)
    775             # Subtle.  Same as in the big comment below.

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in save_cell(pickler, obj)
   1146     f = obj.cell_contents
-> 1147     pickler.save_reduce(_create_cell, (f,), obj=obj)
   1148     log.info("# Ce")

/databricks/python/lib/python3.7/pickle.py in save_reduce(self, func, args, state, listitems, dictitems, obj)
    637             save(func)
--> 638             save(args)
    639             write(REDUCE)

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return

/databricks/python/lib/python3.7/pickle.py in save_tuple(self, obj)
    773             for element in obj:
--> 774                 save(element)
    775             # Subtle.  Same as in the big comment below.

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in save_function(pickler, obj)
   1412                                 obj.__defaults__, obj.__closure__,
-> 1413                                 obj.__dict__, fkwdefaults), obj=obj)
   1414         else:

/databricks/python/lib/python3.7/pickle.py in save_reduce(self, func, args, state, listitems, dictitems, obj)
    637             save(func)
--> 638             save(args)
    639             write(REDUCE)

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return

/databricks/python/lib/python3.7/pickle.py in save_tuple(self, obj)
    788         for element in obj:
--> 789             save(element)
    790 

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    548         # Save the reduce() output and finally memoize the object
--> 549         self.save_reduce(obj=obj, *rv)
    550 

/databricks/python/lib/python3.7/pickle.py in save_reduce(self, func, args, state, listitems, dictitems, obj)
    658         if dictitems is not None:
--> 659             self._batch_setitems(dictitems)
    660 

/databricks/python/lib/python3.7/pickle.py in _batch_setitems(self, items)
    884                     save(k)
--> 885                     save(v)
    886                 write(SETITEMS)

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    523             if reduce is not None:
--> 524                 rv = reduce(self.proto)
    525             else:

/databricks/spark/python/pyspark/context.py in __getnewargs__(self)
    347         raise Exception(
--> 348             "It appears that you are attempting to reference SparkContext from a broadcast "
    349             "variable, action, or transformation. SparkContext can only be used on the driver, "

Exception: It appears that you are attempting to reference SparkContext from a broadcast variable, action, or transformation. SparkContext can only be used on the driver, not in code that it run on workers. For more information, see SPARK-5063.

During handling of the above exception, another exception occurred:

Exception                                 Traceback (most recent call last)
<command-2494588> in <module>
      2 temp = tempfile.gettempdir()
      3 
----> 4 transform_data(train, test, temp)
      5 results = train_and_evaluate(temp)
      6 pprint.pprint(results)

<command-2494583> in transform_data(train_data_file, test_data_file, working_dir)
     42           | 'FixCommasTrainData' >> beam.Map(
     43               lambda line: line.replace(', ', ','))
---> 44           | 'DecodeTrainData' >> MapAndFilterErrors(converter.decode))
     45 
     46       # Combine data and schema into a dataset tuple.  Note that we already used

/databricks/python/lib/python3.7/site-packages/apache_beam/transforms/core.py in Map(fn, *args, **kwargs)
   1524   # pylint: enable=protected-access
   1525 
-> 1526   pardo = FlatMap(wrapper, *args, **kwargs)
   1527   pardo.label = label
   1528   return pardo

/databricks/python/lib/python3.7/site-packages/apache_beam/transforms/core.py in FlatMap(fn, *args, **kwargs)
   1470         'Received %r instead.' % (fn))
   1471 
-> 1472   pardo = ParDo(CallableWrapperDoFn(fn), *args, **kwargs)
   1473   pardo.label = label
   1474   return pardo

/databricks/python/lib/python3.7/site-packages/apache_beam/transforms/core.py in __init__(self, fn, *args, **kwargs)
   1226   """
   1227   def __init__(self, fn, *args, **kwargs):
-> 1228     super(ParDo, self).__init__(fn, *args, **kwargs)
   1229     # TODO(robertwb): Change all uses of the dofn attribute to use fn instead.
   1230     self.dofn = self.fn

/databricks/python/lib/python3.7/site-packages/apache_beam/transforms/ptransform.py in __init__(self, fn, *args, **kwargs)
    786     # Ensure fn and side inputs are picklable for remote execution.
    787     try:
--> 788       self.fn = pickler.loads(pickler.dumps(self.fn))
    789     except RuntimeError as e:
    790       raise RuntimeError('Unable to pickle fn %s: %s' % (self.fn, e))

/databricks/python/lib/python3.7/site-packages/apache_beam/internal/pickler.py in dumps(o, enable_trace)
    256       if enable_trace:
    257         dill.dill._trace(True)  # pylint: disable=protected-access
--> 258         s = dill.dumps(o)
    259       else:
    260         raise

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in dumps(obj, protocol, byref, fmode, recurse, **kwds)
    263     """pickle an object to a string"""
    264     file = StringIO()
--> 265     dump(obj, file, protocol, byref, fmode, recurse, **kwds)#, strictio)
    266     return file.getvalue()
    267 

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in dump(obj, file, protocol, byref, fmode, recurse, **kwds)
    257     _kwds = kwds.copy()
    258     _kwds.update(dict(byref=byref, fmode=fmode, recurse=recurse))
--> 259     Pickler(file, protocol, **_kwds).dump(obj)
    260     return
    261 

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in dump(self, obj)
    443             raise PicklingError(msg)
    444         else:
--> 445             StockPickler.dump(self, obj)
    446         stack.clear()  # clear record of 'recursion-sensitive' pickled objects
    447         return

/databricks/python/lib/python3.7/pickle.py in dump(self, obj)
    435         if self.proto >= 4:
    436             self.framer.start_framing()
--> 437         self.save(obj)
    438         self.write(STOP)
    439         self.framer.end_framing()

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    547 
    548         # Save the reduce() output and finally memoize the object
--> 549         self.save_reduce(obj=obj, *rv)
    550 
    551     def persistent_id(self, obj):

/databricks/python/lib/python3.7/pickle.py in save_reduce(self, func, args, state, listitems, dictitems, obj)
    660 
    661         if state is not None:
--> 662             save(state)
    663             write(BUILD)
    664 

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

/databricks/python/lib/python3.7/site-packages/apache_beam/internal/pickler.py in new_save_module_dict(pickler, obj)
    219         return old_save_module_dict(pickler, obj)
    220     else:
--> 221       return old_save_module_dict(pickler, obj)
    222 
    223   dill.dill.save_module_dict = new_save_module_dict

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in save_module_dict(pickler, obj)
    910             # we only care about session the first pass thru
    911             pickler._session = False
--> 912         StockPickler.save_dict(pickler, obj)
    913         log.info("# D2")
    914     return

/databricks/python/lib/python3.7/pickle.py in save_dict(self, obj)
    857 
    858         self.memoize(obj)
--> 859         self._batch_setitems(obj.items())
    860 
    861     dispatch[dict] = save_dict

/databricks/python/lib/python3.7/pickle.py in _batch_setitems(self, items)
    883                 for k, v in tmp:
    884                     save(k)
--> 885                     save(v)
    886                 write(SETITEMS)
    887             elif n:

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in save_function(pickler, obj)
   1411                                 globs, obj.__name__,
   1412                                 obj.__defaults__, obj.__closure__,
-> 1413                                 obj.__dict__, fkwdefaults), obj=obj)
   1414         else:
   1415             _super = ('super' in getattr(obj.func_code,'co_names',())) and (_byref is not None) and getattr(pickler, '_recurse', False)

/databricks/python/lib/python3.7/pickle.py in save_reduce(self, func, args, state, listitems, dictitems, obj)
    636         else:
    637             save(func)
--> 638             save(args)
    639             write(REDUCE)
    640 

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

/databricks/python/lib/python3.7/pickle.py in save_tuple(self, obj)
    787         write(MARK)
    788         for element in obj:
--> 789             save(element)
    790 
    791         if id(obj) in memo:

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

/databricks/python/lib/python3.7/pickle.py in save_tuple(self, obj)
    772         if n <= 3 and self.proto >= 2:
    773             for element in obj:
--> 774                 save(element)
    775             # Subtle.  Same as in the big comment below.
    776             if id(obj) in memo:

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in save_cell(pickler, obj)
   1145     log.info("Ce: %s" % obj)
   1146     f = obj.cell_contents
-> 1147     pickler.save_reduce(_create_cell, (f,), obj=obj)
   1148     log.info("# Ce")
   1149     return

/databricks/python/lib/python3.7/pickle.py in save_reduce(self, func, args, state, listitems, dictitems, obj)
    636         else:
    637             save(func)
--> 638             save(args)
    639             write(REDUCE)
    640 

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

/databricks/python/lib/python3.7/pickle.py in save_tuple(self, obj)
    772         if n <= 3 and self.proto >= 2:
    773             for element in obj:
--> 774                 save(element)
    775             # Subtle.  Same as in the big comment below.
    776             if id(obj) in memo:

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

/databricks/python/lib/python3.7/site-packages/dill/_dill.py in save_function(pickler, obj)
   1411                                 globs, obj.__name__,
   1412                                 obj.__defaults__, obj.__closure__,
-> 1413                                 obj.__dict__, fkwdefaults), obj=obj)
   1414         else:
   1415             _super = ('super' in getattr(obj.func_code,'co_names',())) and (_byref is not None) and getattr(pickler, '_recurse', False)

/databricks/python/lib/python3.7/pickle.py in save_reduce(self, func, args, state, listitems, dictitems, obj)
    636         else:
    637             save(func)
--> 638             save(args)
    639             write(REDUCE)
    640 

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    502         f = self.dispatch.get(t)
    503         if f is not None:
--> 504             f(self, obj) # Call unbound method with explicit self
    505             return
    506 

/databricks/python/lib/python3.7/pickle.py in save_tuple(self, obj)
    787         write(MARK)
    788         for element in obj:
--> 789             save(element)
    790 
    791         if id(obj) in memo:

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    547 
    548         # Save the reduce() output and finally memoize the object
--> 549         self.save_reduce(obj=obj, *rv)
    550 
    551     def persistent_id(self, obj):

/databricks/python/lib/python3.7/pickle.py in save_reduce(self, func, args, state, listitems, dictitems, obj)
    657 
    658         if dictitems is not None:
--> 659             self._batch_setitems(dictitems)
    660 
    661         if state is not None:

/databricks/python/lib/python3.7/pickle.py in _batch_setitems(self, items)
    883                 for k, v in tmp:
    884                     save(k)
--> 885                     save(v)
    886                 write(SETITEMS)
    887             elif n:

/databricks/python/lib/python3.7/pickle.py in save(self, obj, save_persistent_id)
    522             reduce = getattr(obj, "__reduce_ex__", None)
    523             if reduce is not None:
--> 524                 rv = reduce(self.proto)
    525             else:
    526                 reduce = getattr(obj, "__reduce__", None)

/databricks/spark/python/pyspark/context.py in __getnewargs__(self)
    346         # This method is called when attempting to pickle SparkContext, which is always an error:
    347         raise Exception(
--> 348             "It appears that you are attempting to reference SparkContext from a broadcast "
    349             "variable, action, or transformation. SparkContext can only be used on the driver, "
    350             "not in code that it run on workers. For more information, see SPARK-5063."

Exception: It appears that you are attempting to reference SparkContext from a broadcast variable, action, or transformation. SparkContext can only be used on the driver, not in code that it run on workers. For more information, see SPARK-5063.

**Comment by ibzib** [comment]
tl;dr *Beam does not use pyspark*.

1. This issue occurs because you are running the example code in the Databricks environment. In order to execute code remotely, Beam pickles the main session. In the Databricks environment, the main session must implicitly include the pyspark context, which as the error indicates should never be pickled. This error can be avoided by constructing and submitting the pipeline from outside the Databricks environment.

2. To execute Python code (like TFX) your Spark workers must use the Beam [SDK harness](https://beam.apache.org/roadmap/portability/). As far as I am aware there is no documentation yet on how to install/connect to the Beam Python SDK harness on Databricks.

3. Also, the Beam Spark runner does not yet support Spark 3.0: ([BEAM-7093](https://issues.apache.org/jira/browse/BEAM-7093)).

**Comment by umusa** [comment]
TFX is a wonderful tool, and our work depends on it. 
We have no access to other cloud services except databricks, through which we access AWS/EC2.

Are there some work-arounds ? 
I have searched some solutions like 
      https://stackoverflow.com/questions/29815878/how-to-deal-with-error-spark-5063-in-spark
      https://stackoverflow.com/questions/29996427/how-to-solve-spark-5063-in-nested-map-functions
      https://stackoverflow.com/questions/32196339/error-with-training-logistic-regression-model-on-apache-spark-spark-5063

Is it possible to do some manual application code change to get it work? 

thanks

**Comment by ibzib** [comment]
> Are there some work-arounds ?

To clarify my original comment: even if you work around 1., you will still have to deal with 2. if you want to run your pipeline at scale on a Spark cluster.

> I have searched some solutions like...

None of these involve Apache Beam, so none are related to this issue.

> Is it possible to do some manual application code change to get it work?

No. The code is not the problem here. To re-emphasize, Beam Python requires a specially prepared environment and will not work "out of the box" with Databricks.

**Comment by umusa** [comment]
Could you please let me know how TFX work on google cloud? 
Only Apache Beam can support TFX without spark involved on google cloud ? 
thanks

**Comment by ibzib** [comment]
The preferred way to run TFX on Google Cloud is through [Dataflow](https://cloud.google.com/dataflow). I recommend this blog post to better understand how TFX, Beam, and Dataflow work together: https://blog.tensorflow.org/2020/03/tensorflow-extended-tfx-using-apache-beam-large-scale-data-processing.html

**Comment by arghyaganguly** [comment]
Closing this due to inactivity.Please feel free to reopen.
